### PR TITLE
Add stuck-transaction monitor for wallet transactions

### PR DESCRIPTION
**Pull request by mswilkison**
Implements #4173. A broadcast wallet transaction that stays unconfirmed in the mempool locks the wallet's main UTXO and blocks all subsequent wallet transactions until it confirms. Today this is silent — the broadcast path waits only for mempool acceptance, never for confirmation.

## Change

A Node-scoped `transactionMonitor` (`pkg/tbtc/transaction_monitor.go`):

- The shared `walletTransactionExecutor` registers every broadcast transaction with the monitor — immediately after a successful `BroadcastTransaction`, and again once it's known on-chain (idempotent) — so all four wallet-action types are covered even if a confirmation lookup fails.
- A background loop (started in `runCoordinationLayer`) polls `GetTransactionConfirmations` for each tracked tx on a wall-clock cadence. Confirmed → dropped. Unconfirmed past the stuck threshold → emit `stuck_wallet_transactions_total` + a Warn log naming the wallet and txid (once per tx, and **before** any give-up eviction). Never-confirming transactions are evicted after a maximum tracking age; a full tracking table surfaces `unmonitored_wallet_transactions_total`. Each check pass is time-budgeted so one slow chain call cannot stall the rest.

keep-client only exposes the metric + log; alerting/paging is left to the operator's existing Prometheus/Grafana. Additive to the coarse `gated_sweep_stall_seconds` signal — it surfaces the *specific stuck txid* so an operator can act in minutes.

## Threshold is wall-clock

Tracking uses a wall-clock duration (`defaultStuckTransactionThreshold = 6h`), not a block count — no extra RPC per check. (An earlier revision of this PR used a block count; this supersedes it.)

## Context-aware confirmation lookups

`bitcoin.Chain.GetTransactionConfirmations` now takes a `context.Context` so callers can bound and cancel the lookup. The Electrum implementation threads the caller context (combined with the connection-lifetime context) into its retry loop; the other callers pass `context.Background()`, which the combining preserves as their prior behavior.

This lets the monitor's check pass call the lookup **synchronously**, bounded by the per-pass time budget, instead of spawning a per-lookup goroutine guarded by an in-flight dedup map. When the budget expires the call is cancelled and the pass defers the remaining transactions to the next pass, so a slow or hung backend can no longer strand a goroutine or an in-flight map entry beyond a check pass.

Note: Electrum's `GetTransactionConfirmations` also makes an internal `GetLatestBlockHeight` call — a separate `bitcoin.Chain` method that is not context-threaded. That trailing lookup still self-bounds via Electrum's own request-retry timeout (~2 min) rather than cancelling with the caller context, so a pass can overrun its budget by up to that before being caught on the next iteration. Not a hang or a leak.

## Tests

Cover: no alert below the threshold and at the boundary; exactly-once alert past it; untracking on confirmation; alert-before-eviction for a never-confirming transaction; the table-full metric; and a budgeted check pass cancelling a blocked confirmation lookup rather than stalling. `go build` + `go test ./pkg/tbtc/... ./pkg/clientinfo/...` pass.

## Out of scope / follow-ups

- Automated fee-bumping (RBF/CPFP) — Part B of #4171, targeted at FROST/ROAST.
- Persist/rehydrate the tracked set across restarts — #4175.
- Make the thresholds node-configurable — #4176.

## Summary by CodeRabbit

* **New Features**
  * Added background monitoring for broadcast wallet transactions that stay unconfirmed, including operator alerts for “stuck” transactions and capacity-based handling for untracked overflow.
  * Introduced two new performance counters for stuck and unmonitored wallet transactions, enabled when metrics are available.
  * Enabled monitoring across deposit sweeps, redemptions, moving funds, and moved-funds sweeps.

* **Tests**
  * Added a dedicated transaction-monitor test suite (alerts, eviction, budgeted checks, capacity limits, snapshot ordering).
  * Updated existing action tests to match the new monitoring wiring.

### CODE REVIEW

File: pkg/tbtc/wallet.go [hunk 1]
```diff
@@ -442,6 +455,14 @@ func (wte *walletTransactionExecutor) broadcastTransaction(
 			}
 
 			broadcastTxLogger.Infof("transaction is known on Bitcoin chain")
+
+			if wte.transactionMonitor != nil {
```
Comment (lrsaturnino) [inline pkg/tbtc/wallet.go]: `track()` here only runs after `GetTransactionConfirmations` (line 447) succeeds. If that call keeps failing, line 454 `continue`s and this transaction is never registered — even though `BroadcastTransaction` at line 420 may have already succeeded. Two paths need to register the transaction: right after a successful broadcast (don't wait on the confirmation lookup for that case), and also here when a later lookup finds the transaction known even if this operator's own broadcast call errored. `track()` is already safe to call more than once on the same hash, so both call sites can register it without special-casing.

File: pkg/tbtc/transaction_monitor.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+		}
+
+		// Still unconfirmed (in the mempool) or not found. A lookup error is
+		// treated the same as unconfirmed: it does not indicate the transaction
+		// confirmed and may be transient.
+		outstanding := now.Sub(t.broadcastAt)
+
+		// Give up on transactions that have been unconfirmed for too long (e.g.
+		// dropped from the mempool) so they cannot fill the tracking table.
+		if outstanding > transactionMonitorMaxTrackingAge {
```
Comment (lrsaturnino) [inline pkg/tbtc/transaction_monitor.go:284]: The 24-hour give-up check here runs and `continue`s before the 6-hour stuck-alert check at line 185, in the same pass. If a transaction's first `check()` happens after 24 hours already passed, it gets evicted with only a "giving up" log — the stuck alert and the `stuck_wallet_transactions_total` metric never fire for it. At the defaults, each Electrum operation has a 30-second attempt timeout and a two-minute retry window, but `GetTransactionConfirmations` can perform several such operations. Reaching eviction before the first alert still requires substantial accumulated delay across the serialized checks, or a long process-level pause, so this is a defensive robustness fix rather than a normal-operation blocker. Still worth fixing: swap the order so the alert check runs before eviction.

File: pkg/tbtc/transaction_monitor.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	// Snapshot the tracked set so chain calls are not made under the lock.
+	tm.mu.Lock()
+	snapshot := make(map[bitcoin.Hash]trackedTransaction, len(tm.tracked))
+	for hash, t := range tm.tracked {
+		snapshot[hash] = *t
+	}
+	tm.mu.Unlock()
+
+	for txHash, t := range snapshot {
+		confirmations, err := tm.btcChain.GetTransactionConfirmations(txHash)
```
Comment (lrsaturnino) [inline pkg/tbtc/transaction_monitor.go]: Although Electrum bounds each individual operation by default, checks here are serialized, so one slow call delays every transaction behind it in the same pass. Strongest fix is a monitor-specific end-to-end timeout budget for the whole check. Bounded concurrency would help too, though the Electrum client's shared connection mutex means the underlying wire requests stay serialized either way.

File: pkg/tbtc/transaction_monitor.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+// or CPFP) promptly.
+//
+// The monitor only emits a metric and a warn-level log identifying the wallet
+// and transaction; automated recovery (fee-bumping / RBF) is intentionally out
+// of scope (see threshold-network/keep-core#4171). Alerting is left to the
+// operator's monitoring stack. Every wallet operator that broadcasts a given
+// transaction tracks it independently, so the metric and log are emitted per
+// operator and should be de-duplicated by transaction hash downstream.
+//
+// The tracked set is in-memory only. A transaction that is already stuck when
```
Comment (lrsaturnino) [inline pkg/tbtc/transaction_monitor.go:96]: Already called out in the comment, so not a new finding — flagging because restarts (deploys, upgrades, crashes) aren't rare, so this is a monitoring gap on a fairly regular cadence, not just an edge case. Recommend filing a follow-up issue to persist the tracked set and rehydrate it on startup before this ships to mainnet.

File: pkg/tbtc/transaction_monitor.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	walletPublicKeyHash [20]byte,
+) {
+	tm.mu.Lock()
+	defer tm.mu.Unlock()
+
+	if _, ok := tm.tracked[txHash]; ok {
+		return
+	}
+
+	if len(tm.tracked) >= transactionMonitorMaxTracked {
```
Comment (lrsaturnino) [inline pkg/tbtc/transaction_monitor.go:144]: When the tracking table is full, this just logs a warning and returns — no metric, and the caller in wallet.go has no way to know registration failed. A full table means a real broadcast transaction goes unmonitored, so suggest adding a dedicated metric here, not just a log line.

File: pkg/tbtc/transaction_monitor.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+	"github.com/keep-network/keep-core/pkg/clientinfo"
+)
+
+const (
+	// defaultStuckTransactionThreshold is how long a broadcast wallet
+	// transaction may remain unconfirmed before it is considered stuck and an
+	// alert is raised (~6 hours).
+	defaultStuckTransactionThreshold = 6 * time.Hour
```
Comment (lrsaturnino) [inline pkg/tbtc/transaction_monitor.go:17]: The PR description's "Open for review" section mentions `defaultStuckTransactionThresholdBlocks = 36`, but the constant actually implemented here is `defaultStuckTransactionThreshold = 6 * time.Hour` — a wall-clock duration, not a block count, and a different name. Worth updating the description to match, or confirming wall-clock is the intended final approach (tracking by block height would need an extra RPC per check). On the config question you raised: since none of these four constants can be tuned without a code change today, recommend moving them into node config with the current values as defaults.

File: pkg/tbtc/transaction_monitor_test.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		return recorder.GetCounterValue(clientinfo.MetricStuckWalletTransactionsTotal)
+	}
+
+	// Fresh: not yet stuck.
+	monitor.check()
+	if got := stuckCount(); got != 0 {
+		t.Fatalf("expected no alert for a fresh transaction; got counter [%v]", got)
+	}
+
+	// Past the threshold: flagged as stuck exactly once across repeated checks.
```
Comment (lrsaturnino) [inline pkg/tbtc/transaction_monitor_test.go:146]: One more small description/test mismatch: the body says this covers "no alert at the threshold," but this test checks a fresh transaction and then jumps to `threshold + 1 minute` — the exact boundary itself isn't tested. Please add a case at exactly `defaultStuckTransactionThreshold`, or update that line in the description.

**Review by lrsaturnino — DISMISSED** [review]
LGTM

**Review by piotr-roslaniec — DISMISSED** [review]

**Review by lrsaturnino — DISMISSED** [review]
LGTM

**Comment by piotr-roslaniec** [comment]
This one's already approved by @lrsaturnino — the only thing blocking it now is a conflict with `main` across 9 files (`pkg/tbtc/node.go`, `pkg/tbtc/moving_funds.go`, `pkg/maintainer/spv/spv.go`, `pkg/tbtcpg/deposit_sweep.go`, and a few test files). Could you rebase when you get a chance? Since this is from your fork, only you can push the rebase — should be a quick merge once that's done.

**Review by piotr-roslaniec — APPROVED** [review]
